Log to system console in AudioWorklet processor if setting enabled

ptomato opened the pull request:
#### 7982fa5b773723aa51a7d3879a75f3290aac6a96
<pre>
Log to system console in AudioWorklet processor if setting enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=282269">https://bugs.webkit.org/show_bug.cgi?id=282269</a>

Reviewed by Alex Christensen.

Duplicates console.log messages and exception messages in Worklet global
scopes to the system console if the relevant WebSetting is enabled, very
similarly to how it is done in Worklet global scopes.

* Source/WebCore/worklets/WorkletGlobalScope.cpp:
(WebCore::WorkletGlobalScope::logExceptionToConsole): check relevant
setting.
(WebCore::WorkletGlobalScope::addConsoleMessage): check relevant
setting.

Canonical link: <a href="https://commits.webkit.org/286148@main">https://commits.webkit.org/286148@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb5e69c561e979a6ae50407cfa9299dfe7956c0f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74872 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54302 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79300 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26111 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76989 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63438 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2087 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58804 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17149 "1 flakes 22 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77939 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48959 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64335 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39197 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46241 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21833 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24443 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67402 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22177 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80786 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2190 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1322 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67058 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2339 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64353 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66353 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/16496 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10302 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8450 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11569 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2155 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4943 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2183 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2190 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/worklets/WorkletGlobalScope.cpp:
```diff
@@ -126,13 +126,17 @@ URL WorkletGlobalScope::completeURL(const String& url, ForceUTF8) const
 void WorkletGlobalScope::addConsoleMessage(std::unique_ptr<Inspector::ConsoleMessage>&& message)
 {
+    if (UNLIKELY(settingsValues().logsPageMessagesToSystemConsoleEnabled && message))
+        PageConsoleClient::logMessageToSystemConsole(*message);
```
Can we check sessionID like we do in WorkerGlobalScope::addConsoleMessage?

ptomato replied:
Sure.

ptomato replied:
On second thought, that seems to always suppress the console messages. Is it possible that WorkletGlobalScope sessions are always considered ephemeral?

youennf replied:
I do not think so, SessionID should be process wide and should not be ephemeral for regular web pages.

achristensen07 commented on Source/WebCore/worklets/WorkletGlobalScope.cpp:
```diff
@@ -126,13 +126,19 @@ URL WorkletGlobalScope::completeURL(const String& url, ForceUTF8) const
 void WorkletGlobalScope::logExceptionToConsole(const String& errorMessage, const String& sourceURL, int lineNumber, int columnNumber, RefPtr<ScriptCallStack>&& stack)
 {
+    if (UNLIKELY(settingsValues().logsPageMessagesToSystemConsoleEnabled))
+        PageConsoleClient::logMessageToSystemConsole(*makeUnique<Inspector::ConsoleMessage>(MessageSource::JS, MessageType::Log, MessageLevel::Error, errorMessage, sourceURL, lineNumber, columnNumber));
```
Why do we need a std::unique_ptr?  Can't we just make a temporary ConsoleMessage on the stack instead of allocating just to immediately discard?

Also, I'm not sure if PageConsoleClient::logMessageToSystemConsole is thread safe.  It might be.  I haven't checked.

ptomato replied:
I'm not 100% confident, but from a quick check, the things it uses are `StringBuilder` and `WTFLogAlways()`, which I'd assume are thread safe.

youennf replied:
This should be safe as it is used for workers as well.

achristensen07 commented on Source/WebCore/worklets/WorkletGlobalScope.cpp:
```diff
@@ -126,13 +126,21 @@ URL WorkletGlobalScope::completeURL(const String& url, ForceUTF8) const
 void WorkletGlobalScope::logExceptionToConsole(const String& errorMessage, const String& sourceURL, int lineNumber, int columnNumber, RefPtr<ScriptCallStack>&& stack)
 {
+    if (UNLIKELY(settingsValues().logsPageMessagesToSystemConsoleEnabled)) {
+        Inspector::ConsoleMessage message { MessageSource::JS, MessageType::Log, MessageLevel::Error, errorMessage, *stack };
```
null check stack?

ptomato replied:
I guess we should use the `sourceURL`, `lineNumber`, and `columnNumber` to construct the message instead if it's null. This results in a bit of an awkward `if` and awkward casts from `int` to `unsigned`, but I think it's correct. (Not sure where the `int`/`unsigned` mismatch originates from, but I'd call it out of scope for this PR.)

ptomato committed:
Log to system console in AudioWorklet processor if setting enabled
https://bugs.webkit.org/show_bug.cgi?id=282269

Reviewed by Alex Christensen.

Duplicates console.log messages and exception messages in Worklet global
scopes to the system console if the relevant WebSetting is enabled, very
similarly to how it is done in Worklet global scopes.

* Source/WebCore/worklets/WorkletGlobalScope.cpp:
(WebCore::WorkletGlobalScope::logExceptionToConsole): check relevant
setting.
(WebCore::WorkletGlobalScope::addConsoleMessage): check relevant
setting.

Canonical link: https://commits.webkit.org/286148@main
